lib/ae_mdw_web/websocket/subscriptions.ex

```diff
@@ -48,41 +50,43 @@ defmodule AeMdwWeb.Websocket.Subscriptions do
     {:ok, :no_state}
   end
 
-  @spec subscribe(pid(), channel()) ::
+  @spec subscribe(pid(), version(), channel()) ::
           {:ok, [channel()]} | {:error, :invalid_channel | :already_subscribed}
-  def subscribe(pid, channel) do
-    with {:ok, channel_key} <- validate_subscribe(pid, channel) do
+  def subscribe(pid, version, channel) do
+    with {:ok, channel_key} <- validate_subscribe(pid, version, channel) do
       maybe_monitor(pid)
-      :ets.insert(@subs_pid_channel, {pid, channel})
+      :ets.insert(@subs_pid_channel, {pid, {channel_key, channel}})
       :ets.insert(@subs_channel_pid, {channel_key, pid})
 
       {:ok, subscribed_channels(pid)}
     end
   end
 
-  @spec unsubscribe(pid(), channel()) ::
+  @spec unsubscribe(pid(), version(), channel()) ::
           {:ok, [channel()]} | {:error, :invalid_channel | :not_subscribed}
-  def unsubscribe(pid, channel) do
-    with {:ok, channel_key} <- validate_unsubscribe(pid, channel) do
+  def unsubscribe(pid, version, channel) do
+    with {:ok, channel_key, channel} <- validate_unsubscribe(pid, version, channel) do
```

Comment:
The channel returned by `validate_unsubscribe` is the same value as the parameter (additional out = in).

Reply:
The channel out value is not always the same as the in value, since it will always return the value of the first subscription

Reply:
There are only two representations for a channel:
A) encoded String, received as a parameter on `unsubscribe` and passed to `validate_unsubscribe`
B) as a raw binary, that is a channel key

Since it already returns the channel key and `validate_unsubscribe` receives the channel as a parameter there is no need to add it to the return of `validate_unsubscribe`. The `validate_subscribe` can be improved to avoid double subscriptions for same object `ak_wTPFpksUJFjjntonTvwK4LJvDw11DPma7kZBneKbumb8yPeFq` and `ok_wTPFpksUJFjjntonTvwK4LJvDw11DPma7kZBneKbumb8yPeFq` (until `ok_wTPFpksUJFjjntonTvwK4LJvDw11DPma7kZBneKbumb8yPeFq` is handled only for Oracles which should be part of another PR please if you wanna work on that right away)

Reply:
The subs_pid_channel table stores values such as:
```
{{<#PID1>, {:v1, <<1, 2, 3, ..>>}}, "ak_123.."} 
```
since the key is the binary representation you can't really subscribe to the same pubkey twice, it will return an error.

When unsubscribing from the record, though, I need both the binary and encoded representation, which is why `validate_unsubscribe` returns the original encoded address the user subscribed to

If this is not what you had in mind, would you mind posting the full-code of what you would like it to be? I can't seem to understand otherwise

Reply:
> The channel out value is not always the same as the in value, since it will always return the value of the first subscription

It was mentioned that it would return the first but the `validate_subscribe` on [master](https://github.com/aeternity/ae_mdw/blob/7d671009dd6d58cf658eedef41210edbf9ecbe7b/lib/ae_mdw_web/websocket/subscriptions.ex#L140-L142) already doesn't allow `ok_wTPFpksUJFjjntonTvwK4LJvDw11DPma7kZBneKbumb8yPeFq` to be subscribed when `ak_wTPFpksUJFjjntonTvwK4LJvDw11DPma7kZBneKbumb8yPeFq` already is. 

So the first entry would be the only subscribed channel available as checked by:
https://github.com/aeternity/ae_mdw/blob/7d671009dd6d58cf658eedef41210edbf9ecbe7b/lib/ae_mdw_web/websocket/subscriptions.ex#L164

If the channel subscribed was `ak_wTPFpksUJFjjntonTvwK4LJvDw11DPma7kZBneKbumb8yPeFq` I think it's fair and expected that the `unsubscribe` works only with same string and not `ok_wTPFpksUJFjjntonTvwK4LJvDw11DPma7kZBneKbumb8yPeFq`. Following this behavior, the `validate_unsubscribe` doesn't need to return the `channel` because it's already the input.

Reply:
@sborrazas I understand the confusion that is happening. It was due to the `match` function returning the channel while if `exists?` it just had to detect a match or not. Thanks going through it avoiding the return of the channel. 

Please find here the code requested with minimum changes needed:
https://github.com/aeternity/ae_mdw/commit/b13edb5fb1eac4d9ad4184d23b8261115b78456d
- `validate_unsubscribe` returns only the `channel_key`
- `exists?` depends only channel key
- `subscribed_channels` only flattens the list
- does not allow duplicate subscription of account and oracle with same pubkey (as you've fixed before, added only a test case)
- since oracle key returns `:already_subscribed` for same account pubkey, it allows unsubscribing it (same as above).